Implement adding a rectangle to a pdf layer

dodomorandi opened the pull request:
This should improve the support to writing PDFs towards two directions:

1. The ability of stroke/fill rectangles
2. The possibility of using simple clipping operations based on rectangles

The changes should be pretty straightforward, and I also added a few changes/fixes to the doc. Still, there are a couple of points I would like to discuss:

- I added `Mode` and `Winding` specifically for the `rectangle` module. However, we could consider to unify those with the one available inside the `line` module. The `Winding` enum cannot actually be any different, I am a bit unsure about the `Mode` because I don't have a clear view about other possibilities (i.e: is there a paint mode that we would like to add in the future that makes sense for a `Rect` but not for a `Polygon` or vice versa?)
- While writing the implementation for `into_stream_op`, I noticed that `Polygon` [adds a __paint end__](https://github.com/fschutt/printpdf/blob/cec012ad726344c31f715ab8c939ad0c11c7fcaf/src/line.rs#L291) even after a stroke/fill. AFAIK (reading from the PDF specs) this should be only necessary if a _path-painting operator_ has not been used -- and strokes/fills are actually _path-painting operators_. I am pointing this out just because if I am right, I don't think it should be an actual problem (probably any pdf reader does not have any trouble ignoring the `n` operator), but if I am wrong I should change the implementation for `Rect`, because I only add the _path end operator_ only in clipping mode.

Let me know what do you think, and feel free to give any suggestion I could improve the code or the doc.

Last but not least: this is a breaking change because of the new `Rect::mode` field (and [cargo-semver-checks](https://github.com/obi1kenobi/cargo-semver-checks) confirms that).

CC @paolobarbolini

dodomorandi committed:
doc: add missing docs for rectangle module

dodomorandi committed:
doc: fix doc in rectangle module

dodomorandi committed:
feat!: impl adding rectangles in pdf layers

With this implementation is also possible to use rectangles to clip
other paths.

dodomorandi committed:
chore: add example for using a rectangle

fschutt commented on src/rectangle.rs:
```diff
@@ -32,3 +89,46 @@ impl PartialEq for Rect {
+
+/// The paint mode of a [`Rect`].
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Mode {
```
Use the existing `PolygonMode` and `WindingOrder` enums.

dodomorandi replied:
Do you want me to leave `PolygonMode` and `WindingOrder` in the current module? As I pointed out before, it would make sense to have a unique version, but on the other hand it would make sense to move to a _common_ module and, in the case of the `PolygonMode`, rename it IMHO.

fschutt replied:
yeah, move it out if it's shared between modules

fschutt commented on src/rectangle.rs:
```diff
@@ -32,3 +89,46 @@ impl PartialEq for Rect {
+    FillStroke,
+
+    /// The shape is not painted, and it is used to clip other paths.
+    Clip(
```
You could also have a `FillStroke` with a `NonZero` and `EvenOdd` winding, so only adding `Winding` on the `Clip` is wrong, mode and winding are completely separate and can be combined in creative ways.

dodomorandi replied:
You are right, but at that point I need to introduce all the possible combinations, which are not currently supported for the polygon as well.

fschutt commented on src/rectangle.rs:
```diff
@@ -32,3 +89,46 @@ impl PartialEq for Rect {
+
+/// The clipping path operator.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Winding {
```
Use `PolygonWinding`, already exists.

fschutt commented on src/rectangle.rs:
```diff
@@ -1,15 +1,26 @@
 #[derive(Debug, Copy, Clone)]
 pub struct Rect {
-    /// x position from the bottom left corner in pt
+    /// Position of the lower left point of the rectangle, relative to the bottom left corner.
```
I'd add "bottom left corner of the PDF page" and "in point" at the end so it's immediately clear when reading the documentation.

dodomorandi replied:
Ok, just keep in mind that this is not an invariance provided by `Rect`, but it is related to how `Point` works. If you mean to change how `Point` work then you need to remember to update this documentation.

fschutt commented on src/pdf_layer.rs:
```diff
@@ -548,5 +548,11 @@ impl PdfLayerReference {
-}
 
+    /// Add a rectangle to the layer.
+    pub fn add_rectangle(&self, rect: Rect) {
```
I'd rename it `add_rect` so it's not so long and fits in the the `Rect` struct.

fschutt commented on Cargo.toml:
```diff
@@ -108,3 +108,11 @@ required-features = []
 name = "hyperlink"
 required-features = ["annotations"]
+
+[[example]]
```
I'd also rename these examples `rect` and `clipping_rect` respectively.

fschutt commented on examples/clipping_rectangle.rs:
I'd merge them into one example (one rect and one clipping rect on one page).

fschutt reviewed (changes requested):
Some changes, most notably `WindingOrder` and `PolygonMode` already exist

dodomorandi committed:
refactor!: move path-related items to module

This should improve a bit the API once `WindingOrder` and `PaintMode`
are also used by the `Rect`.

dodomorandi committed:
fix: minor clippy lints on path module

dodomorandi committed:
feat: missing derives on items in path module

dodomorandi committed:
doc: add docs to path module

dodomorandi committed:
feat: add WindingOrder::get_fill_stroke_op
